### PR TITLE
Don't search for oneDPL if the compiler already provides it

### PR DESCRIPTION
**Pull request by masterleinad**
In general, `oneAPI` provides `oneDPL` but no `CMake` find package file.
This pull request proposes to skip finding and linking `oneDPL` if the compiler can already include relevant headers.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -31,8 +31,17 @@ endif()
...
+  check_include_file_cxx(oneapi/dpl/execution ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER)
+  check_include_file_cxx(oneapi/dpl/algorithm ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER)
+  if (!ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER OR !ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER)
+    find_package(oneDPL REQUIRED)
+    target_link_libraries(ArborX INTERFACE oneDPL)
+  endif()
+  # libstdc++-9 and libstdc++-10 using legacy TBB that is incompatible with oneTBB, see
+  # https://stackoverflow.com/a/67924408
+  target_compile_definitions(ArborX INTERFACE _GLIBCXX_USE_TBB_PAR_BACKEND=0)
+  target_compile_definitions(ArborX INTERFACE PSTL_USE_PARALLEL_POLICIES=0)
```
Comment (masterleinad) [inline CMakeLists.txt:44]: This is debatable but pretty much everyone using `oneDPL` (with GCC-9 or GCC-10 headers) needs to do this one way or the other.

File: cmake/ArborXConfig.cmake.in [hunk 2]
```diff
@@ -16,7 +16,9 @@ if(Kokkos_ENABLE_HIP AND ARBORX_ENABLE_ROCTHRUST)
   find_dependency(rocthrust)
 endif()
 if(Kokkos_ENABLE_SYCL AND ARBORX_ENABLE_ONEDPL)
-  find_dependency(oneDPL)
+  if(!@ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER@ OR !@ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER@)
```
[the remark this answers is not in the document]
Comment (dalg24) [reply cmake/ArborXConfig.cmake.in:19]: either define a temporary variable (something like `ARBORX_CXX_COMPILER_BUNDLED_ONEDPL`) or add a comment since this will be unreadable after the file is configured.

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (masterleinad) [reply CMakeLists.txt:44]: Obviously, the compiler we use for CI doesn't need it but I need it on the testbeds. It seems to be a toolchain issue depending on the standard library implementation used.
In any case, I dropped it for now. We can discuss adding it elsewhere.

File: CMakeLists.txt [hunk 3]
```diff
@@ -31,8 +31,14 @@ endif()
 
 cmake_dependent_option(ARBORX_ENABLE_ONEDPL "Enable oneDPL support" ON "Kokkos_ENABLE_SYCL" OFF)
 if(Kokkos_ENABLE_SYCL AND ARBORX_ENABLE_ONEDPL)
-  find_package(oneDPL REQUIRED)
-  target_link_libraries(ArborX INTERFACE oneDPL)
+  include(CheckIncludeFileCXX)
+  check_include_file_cxx(oneapi/dpl/execution ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER)
+  check_include_file_cxx(oneapi/dpl/algorithm ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER)
+  if (NOT ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER OR
+      NOT ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER)
```
Comment (dalg24) [inline CMakeLists.txt:38]:
```suggestion
  set(ARBORX_CXX_COMPILER_BUNDLED_ONEDPL
      ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER AND
      ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER)
  if(NOT ARBORX_CXX_COMPILER_BUNDLED_ONEDPL)
```

File: cmake/ArborXConfig.cmake.in [hunk 4]
```diff
@@ -16,7 +16,12 @@ if(Kokkos_ENABLE_HIP AND ARBORX_ENABLE_ROCTHRUST)
   find_dependency(rocthrust)
 endif()
 if(Kokkos_ENABLE_SYCL AND ARBORX_ENABLE_ONEDPL)
-  find_dependency(oneDPL)
+  # NOT @ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER@ OR
+  # NOT @ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER@
+  if(NOT @ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER@ OR
+     NOT @ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER@)
+    find_dependency(oneDPL)
+  endif()
```
Comment (dalg24) [inline cmake/ArborXConfig.cmake.in:24]:
```suggestion
  set(ARBORX_CXX_COMPILER_BUNDLED_ONEDPL @ARBORX_CXX_COMPILER_BUNDLED_ONEDPL@)
  if(NOT ARBORX_CXX_COMPILER_BUNDLED_ONEDPL)
    find_dependency(oneDPL)
  endif()
  unset(ARBORX_CXX_COMPILER_BUNDLED_ONEDPL)
```

**Review by dalg24 — COMMENTED** [review]
I think it is easier to read with the changes I suggested

File: cmake/ArborXSettings.cmake.in [hunk 5]
```diff
@@ -1,4 +1,6 @@
 set(ARBORX_ENABLE_ROCTHRUST @ARBORX_ENABLE_ROCTHRUST@)
 set(ARBORX_ENABLE_ONEDPL @ARBORX_ENABLE_ONEDPL@)
+set(ARBORX_CXX_COMPILER_BUNDLED_ONEDPL @ARBORX_CXX_COMPILER_BUNDLED_ONEDPL@)
```
Comment (dalg24) [inline cmake/ArborXSettings.cmake.in:3]: Why did you move it?

Comment (masterleinad) [reply cmake/ArborXSettings.cmake.in:3]: Because we have all the other configuration variables here but I don't care much. Would rather have it moved back?

Comment (dalg24) [reply cmake/ArborXSettings.cmake.in:3]: The other are actual variables we export.  This one we define to use in the condition before unsetting it.
Please undo that change.

Comment (masterleinad) [reply cmake/ArborXSettings.cmake.in]: Here you go.

File: CMakeLists.txt [hunk 6]
```diff
@@ -31,8 +31,17 @@ endif()
...
 cmake_dependent_option(ARBORX_ENABLE_ONEDPL "Enable oneDPL support" ON "Kokkos_ENABLE_SYCL" OFF)
 if(Kokkos_ENABLE_SYCL AND ARBORX_ENABLE_ONEDPL)
-  find_package(oneDPL REQUIRED)
-  target_link_libraries(ArborX INTERFACE oneDPL)
+  include(CheckIncludeFileCXX)
+  check_include_file_cxx(oneapi/dpl/execution ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER)
+  check_include_file_cxx(oneapi/dpl/algorithm ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER)
+  if (ARBORX_COMPILER_HAS_ONEDPL_EXECUTION_HEADER AND
+      ARBORX_COMPILER_HAS_ONEDPL_ALGORITHM_HEADER)
+    set(ARBORX_CXX_COMPILER_BUNDLED_ONEDPL TRUE)
```
Comment (dalg24) [inline CMakeLists.txt:39]: else set to `FALSE` will probably produce a better configured file

**Comment by masterleinad** [comment]
Retest this please.

File: cmake/ArborXSettings.cmake.in [hunk 7]
```diff
@@ -2,3 +2,4 @@ set(ARBORX_ENABLE_ROCTHRUST @ARBORX_ENABLE_ROCTHRUST@)
 set(ARBORX_ENABLE_ONEDPL @ARBORX_ENABLE_ONEDPL@)
 set(ARBORX_ENABLE_MPI @ARBORX_ENABLE_MPI@)
 set(ARBORX_USE_CUDA_AWARE_MPI @ARBORX_USE_CUDA_AWARE_MPI@)
+
```
Comment (aprokop) [inline cmake/ArborXSettings.cmake.in:5]: Would be nice to not have this new line change.

**Review by aprokop — APPROVED** [review]

**Comment by aprokop** [comment]
Does ArborX with this patch build on Arcticus?

**Comment by masterleinad** [comment]
> @masterleinad Does ArborX with this patch build on Arcticus?

No, you also need something like the code dropped in 
[`bdc69c4` (#709)](https://github.com/arborx/ArborX/pull/709/commits/bdc69c46108d43dd3d5321907f3ae138c5fcfb70)

**Comment by aprokop** [comment]
SYCL test hanged in installation:
```
+ make test
Running tests...
Test project /var/jenkins/workspace/ArborX_PR-709/test_install
    Start 1: ArborX_HostAccessTraits_Example
Sending interrupt signal to process
Sending interrupt signal to process
Terminated
Makefile:102: recipe for target 'test' failed
make: *** [test] Terminated
script returned exit code 143
```

**Comment by masterleinad** [comment]
Retest this please.

**Comment by aprokop** [comment]
The warnings in the build are from the containers. Merging.
